File: java/src/test/java/com/incidentcommander/StateMachineTest.java
package com.incidentcommander;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.Instant;
import org.junit.jupiter.api.Test;

class StateMachineTest {
    private static final Instant EPOCH = Instant.EPOCH;

    private static Incident newTestIncident() {
        Alert alert = new Alert();
        alert.service = Services.DATABASE;
        alert.metric = "error_rate";
        alert.severity = Severity.CRITICAL;
        return StateMachine.newIncident("inc-1", alert, EPOCH);
    }

    @Test
    void newStartsInNew() {
        Incident inc = newTestIncident();
        assertEquals(States.NEW, inc.state);
        assertEquals(1, inc.history.size());
        assertEquals(States.NEW, inc.history.get(0).to);
        assertEquals("", inc.history.get(0).from);
        assertEquals("incident opened", inc.history.get(0).reason);
    }

    @Test
    void newSetsTitleFromAlert() {
        assertEquals("error_rate on database", newTestIncident().title);
    }

    @Test
    void happyPathWalksToResolved() {
        Incident inc = newTestIncident();
        String[] path = {
            States.TRIAGING,
            States.INVESTIGATING,
            States.ROOT_CAUSE_IDENTIFIED,
            States.REMEDIATION_PROPOSED,
            States.WAITING_FOR_APPROVAL,
            States.REMEDIATING,
            States.VERIFYING,
            States.RESOLVED,
        };
        for (String to : path) {
            StateMachine.transition(inc, to, "step", EPOCH);
        }
        assertEquals(States.RESOLVED, inc.state);
        assertEquals(9, inc.history.size());
    }

    @Test
    void illegalTransitionIsRejectedAndDoesNotMutate() {
        Incident inc = newTestIncident();
        IllegalTransitionException ex = assertThrows(
                IllegalTransitionException.class,
                () -> StateMachine.transition(inc, States.REMEDIATING, "skip", EPOCH));
        assertEquals(States.NEW, ex.getFrom());
        assertEquals(States.REMEDIATING, ex.getTo());
        assertEquals(States.NEW, inc.state);
        assertEquals(1, inc.history.size());
    }

    @Test
    void illegalTransitionMessage() {
        Incident inc = newTestIncident();
        IllegalTransitionException ex = assertThrows(
                IllegalTransitionException.class,
                () -> StateMachine.transition(inc, States.RESOLVED, "x", EPOCH));
        assertEquals("illegal transition NEW -> RESOLVED", ex.getMessage());
    }

    @Test
    void terminalStatesHaveNoExit() {
        String[] terminals = {
            States.RESOLVED,
            States.ESCALATED,
            States.INVESTIGATION_FAILED,
            States.REMEDIATION_FAILED,
            States.VERIFICATION_FAILED,
        };
        for (String state : terminals) {
            assertTrue(StateMachine.isTerminal(state));
            Incident inc = newTestIncident();
            inc.state = state;
            assertThrows(
                    IllegalTransitionException.class,
                    () -> StateMachine.transition(inc, States.RESOLVED, "x", EPOCH));
        }
    }

    @Test
    void nonTerminalStatesAreNotTerminal() {
        String[] states = {
            States.NEW,
            States.TRIAGING,
            States.INVESTIGATING,
            States.ROOT_CAUSE_IDENTIFIED,
            States.REMEDIATION_PROPOSED,
            States.WAITING_FOR_APPROVAL,
            States.REMEDIATING,
            States.VERIFYING,
        };
        for (String state : states) {
            assertFalse(StateMachine.isTerminal(state));
        }
    }

    @Test
    void investigationCanFail() {
        Incident inc = newTestIncident();
        StateMachine.transition(inc, States.TRIAGING, "step", EPOCH);
        StateMachine.transition(inc, States.INVESTIGATING, "step", EPOCH);
        StateMachine.transition(inc, States.INVESTIGATION_FAILED, "no evidence", EPOCH);
        assertTrue(StateMachine.isTerminal(inc.state));
    }

    @Test
    void readOnlyResolutionSkipsRemediation() {
        Incident inc = newTestIncident();
        StateMachine.transition(inc, States.TRIAGING, "step", EPOCH);
        StateMachine.transition(inc, States.INVESTIGATING, "step", EPOCH);
        StateMachine.transition(inc, States.ROOT_CAUSE_IDENTIFIED, "step", EPOCH);
        StateMachine.transition(inc, States.RESOLVED, "informational", EPOCH);
        assertEquals(States.RESOLVED, inc.state);
    }

    @Test
    void canTransitionMatchesTable() {
        assertTrue(StateMachine.canTransition(States.NEW, States.TRIAGING));
        assertFalse(StateMachine.canTransition(States.NEW, States.RESOLVED));
        assertTrue(StateMachine.canTransition(States.ROOT_CAUSE_IDENTIFIED, States.ESCALATED));
        assertFalse(StateMachine.canTransition(States.RESOLVED, States.TRIAGING));
    }

    @Test
    void canTransitionUnknownStateIsFalse() {
        assertFalse(StateMachine.canTransition("BOGUS", States.TRIAGING));
    }

    @Test
    void transitionRecordsFromAndReason() {
        Incident inc = newTestIncident();
        StateMachine.transition(inc, States.TRIAGING, "begin triage", EPOCH);
        Transition last = inc.history.get(inc.history.size() - 1);
        assertEquals(States.NEW, last.from);
        assertEquals(States.TRIAGING, last.to);
        assertEquals("begin triage", last.reason);
        assertEquals(EPOCH, inc.updatedAt);
    }
}

File: rust/tests/statemachine_test.rs
use incident_commander::models::{
    Alert, Incident, SERVICE_DATABASE, SEVERITY_CRITICAL, STATE_ESCALATED, STATE_INVESTIGATING,
    STATE_INVESTIGATION_FAILED, STATE_NEW, STATE_REMEDIATING, STATE_REMEDIATION_FAILED,
    STATE_REMEDIATION_PROPOSED, STATE_RESOLVED, STATE_ROOT_CAUSE_IDENTIFIED, STATE_TRIAGING,
    STATE_VERIFICATION_FAILED, STATE_VERIFYING, STATE_WAITING_FOR_APPROVAL,
};
use incident_commander::statemachine::{
    can_transition, is_terminal, new_incident, transition, IllegalTransition,
};

const EPOCH: i64 = 0;

fn new_test_incident() -> Incident {
    let alert = Alert {
        service: SERVICE_DATABASE.to_string(),
        metric: "error_rate".to_string(),
        severity: SEVERITY_CRITICAL.to_string(),
        ..Default::default()
    };
    new_incident("inc-1", alert, EPOCH)
}

#[test]
fn new_starts_in_new() {
    let inc = new_test_incident();
    assert_eq!(inc.state, STATE_NEW);
    assert_eq!(inc.history.len(), 1);
    assert_eq!(inc.history[0].to, STATE_NEW);
    assert_eq!(inc.history[0].from, "");
    assert_eq!(inc.history[0].reason, "incident opened");
}

#[test]
fn new_sets_title_from_alert() {
    assert_eq!(new_test_incident().title, "error_rate on database");
}

#[test]
fn happy_path_walks_to_resolved() {
    let mut inc = new_test_incident();
    let path = [
        STATE_TRIAGING,
        STATE_INVESTIGATING,
        STATE_ROOT_CAUSE_IDENTIFIED,
        STATE_REMEDIATION_PROPOSED,
        STATE_WAITING_FOR_APPROVAL,
        STATE_REMEDIATING,
        STATE_VERIFYING,
        STATE_RESOLVED,
    ];
    for to in path {
        transition(&mut inc, to, "step", EPOCH).unwrap();
    }
    assert_eq!(inc.state, STATE_RESOLVED);
    assert_eq!(inc.history.len(), 9);
}

#[test]
fn illegal_transition_is_rejected_and_does_not_mutate() {
    let mut inc = new_test_incident();
    let err = transition(&mut inc, STATE_REMEDIATING, "skip", EPOCH).unwrap_err();
    assert_eq!(
        err,
        IllegalTransition {
            from: STATE_NEW.to_string(),
            to: STATE_REMEDIATING.to_string(),
        }
    );
    assert_eq!(inc.state, STATE_NEW);
    assert_eq!(inc.history.len(), 1);
}

#[test]
fn illegal_transition_message() {
    let mut inc = new_test_incident();
    let err = transition(&mut inc, STATE_RESOLVED, "x", EPOCH).unwrap_err();
    assert_eq!(err.to_string(), "illegal transition NEW -> RESOLVED");
}

#[test]
fn terminal_states_have_no_exit() {
    for state in [
        STATE_RESOLVED,
        STATE_ESCALATED,
        STATE_INVESTIGATION_FAILED,
        STATE_REMEDIATION_FAILED,
        STATE_VERIFICATION_FAILED,
    ] {
        assert!(is_terminal(state));
        let mut inc = new_test_incident();
        inc.state = state.to_string();
        assert!(transition(&mut inc, STATE_RESOLVED, "x", EPOCH).is_err());
    }
}

#[test]
fn non_terminal_states_are_not_terminal() {
    for state in [
        STATE_NEW,
        STATE_TRIAGING,
        STATE_INVESTIGATING,
        STATE_ROOT_CAUSE_IDENTIFIED,
        STATE_REMEDIATION_PROPOSED,
        STATE_WAITING_FOR_APPROVAL,
        STATE_REMEDIATING,
        STATE_VERIFYING,
    ] {
        assert!(!is_terminal(state));
    }
}

#[test]
fn investigation_can_fail() {
    let mut inc = new_test_incident();
    transition(&mut inc, STATE_TRIAGING, "step", EPOCH).unwrap();
    transition(&mut inc, STATE_INVESTIGATING, "step", EPOCH).unwrap();
    transition(&mut inc, STATE_INVESTIGATION_FAILED, "no evidence", EPOCH).unwrap();
    assert!(is_terminal(&inc.state));
}

#[test]
fn read_only_resolution_skips_remediation() {
    let mut inc = new_test_incident();
    transition(&mut inc, STATE_TRIAGING, "step", EPOCH).unwrap();
    transition(&mut inc, STATE_INVESTIGATING, "step", EPOCH).unwrap();
    transition(&mut inc, STATE_ROOT_CAUSE_IDENTIFIED, "step", EPOCH).unwrap();
    transition(&mut inc, STATE_RESOLVED, "informational", EPOCH).unwrap();
    assert_eq!(inc.state, STATE_RESOLVED);
}

#[test]
fn can_transition_matches_table() {
    assert!(can_transition(STATE_NEW, STATE_TRIAGING));
    assert!(!can_transition(STATE_NEW, STATE_RESOLVED));
    assert!(can_transition(STATE_ROOT_CAUSE_IDENTIFIED, STATE_ESCALATED));
    assert!(!can_transition(STATE_RESOLVED, STATE_TRIAGING));
}

#[test]
fn can_transition_unknown_state_is_false() {
    assert!(!can_transition("BOGUS", STATE_TRIAGING));
}

#[test]
fn transition_records_from_and_reason() {
    let mut inc = new_test_incident();
    transition(&mut inc, STATE_TRIAGING, "begin triage", EPOCH).unwrap();
    let last = inc.history.last().unwrap();
    assert_eq!(last.from, STATE_NEW);
    assert_eq!(last.to, STATE_TRIAGING);
    assert_eq!(last.reason, "begin triage");
    assert_eq!(inc.updated_at, EPOCH);
}
